src/rebar_prv_install_deps.erl

```diff
@@ -184,33 +153,74 @@ handle_deps(Profile, State, Deps, Upgrade, Locks) ->
 %% Internal functions
 %% ===================================================================
 
+deps_per_profile(Profile, {Apps, State}) ->
+    Locks = rebar_state:get(State, {locks, Profile}, []),
+    ProfileDeps = rebar_state:get(State, {deps, Profile}, []),
+    {ok, NewApps, NewState} = handle_deps(Profile, State, ProfileDeps, Locks),
+    {NewApps++Apps, NewState}.
+
+find_cycles(Apps) ->
+    case rebar_digraph:compile_order(Apps) of
+        {error, {cycles, Cycles}} -> {cycles, Cycles};
+        {error, Error} -> {error, Error};
+        {ok, _} -> no_cycle
+    end.
+
+compile_order(Source, ProjectApps) ->
+    case rebar_digraph:compile_order(Source) of
+        {ok, Sort} ->
+            %% Valid apps are compiled and good
+            {ok, lists:dropwhile(fun rebar_app_info:valid/1, Sort -- ProjectApps)};
+        {error, Error} ->
+            {error, Error}
+    end.
+
+update_pkg_deps(Profile, Packages, PkgDeps, Graph, Upgrade, Seen, State) ->
+    case PkgDeps of
+        [] -> %% No pkg deps
+            {[], State};
+        PkgDeps ->
+            %% Find pkg deps needed
+            S = case rebar_digraph:cull_deps(Graph, PkgDeps) of
+                {ok, [], _} ->
+                    throw({rebar_digraph, no_solution});
+                {ok, Solution, []} ->
+                    Solution;
+                {ok, Solution, Discarded} ->
+                    [warn_skip_pkg(Pkg) || Pkg <- Discarded],
+                    Solution
+            end,
+            update_pkg_deps(Profile, S, Packages, Upgrade, Seen, State)
+    end.
+
 update_pkg_deps(Profile, Pkgs, Packages, Upgrade, Seen, State) ->
     %% Create app_info record for each pkg dep
     DepsDir = rebar_dir:deps_dir(State),
     {Solved, _, State1}
         = lists:foldl(fun(Pkg, {Acc, SeenAcc, StateAcc}) ->
-                              AppInfo = package_to_app(DepsDir
-                                                      ,Packages
-                                                      ,Pkg),
-                              {SeenAcc1, StateAcc1} = maybe_lock(Profile, AppInfo, SeenAcc, StateAcc, 0),
-                              case maybe_fetch(AppInfo, Upgrade, SeenAcc1, State) of
-                                  true ->
-                                      {[AppInfo | Acc], SeenAcc1, StateAcc1};
-                                  false ->
-                                      {Acc, SeenAcc1, StateAcc1}
-                              end
+                        handle_pkg_dep(Profile, Pkg, Packages, Upgrade, DepsDir, Acc, SeenAcc, StateAcc)
```

Comment:
Not all of them get mutated the same way, and some of them need to remain the same no matter 
what.
For example in `handle_seen_src_deps` and `handle_unseen_src_deps`, the former needs both the new locks and the initial locks, whereas the latter doesn't care. Either we'd need to do some manual wrangling for records to make sure they don't get rewritten by the inner function, or it's done explicitly as it is right now.
I have started with a minor refactoring for now, but yeah. It's possible some of these things could be put in a record instead. It'd just mean all the state is passed to all the functions whether they need it or not.

Take the example of calls to `maybe_lock` (https://github.com/rebar/rebar3/pull/171/files#diff-5bbc70cb457999c3dabeeac51e71f8a6R309) where we sometimes override arguments by hand there. With records we will probably still need to do that, but now we will make implicit assumptions from the call-site on how records are read, rather than on function arguments, which is more reasonable.

I'm not sure there is that big of a benefit from using records in these cases as they could lead to making some critical distinctions less explicit than they are right now. I don't mind the pain of writing these functions too much when reading them will be the big issue in the future.